lib/src/widgets/credit_card.dart

```diff
@@ -62,38 +60,67 @@ class _CreditCardState extends State<CreditCard> {
 
     setState(() => isSubmitting = false);
 
+    _handlePaymentResponse(result);
+  }
+
+  void _handlePaymentResponse(result) {
     if (result is! PaymentResponse ||
         result.status != PaymentStatus.initiated) {
       widget.onPaymentResult(result);
       return;
     }
+    if (result.source is CardPaymentResponseSource) {
+      final String transactionUrl =
+          (result.source as CardPaymentResponseSource).transactionUrl;
+      if (mounted) {
+        _openThreeDSecure(
+          result,
+          transactionUrl,
+        );
+      }
+    }
+  }
+
+  void _openThreeDSecure(
+    PaymentResponse result,
+    String transactionUrl,
+  ) {
+    Navigator.push(
+      context,
+      MaterialPageRoute(
+        fullscreenDialog: true,
+        maintainState: false,
+        builder: (context) => ThreeDSWebView(
+          transactionUrl: transactionUrl,
+          callbackUrl: widget.config.callbackUrl,
+          on3dsDone: (
+            String status,
+            String message,
+          ) =>
+              _on3dsDone(
+            result,
+            status,
+            message,
+          ),
+        ),
+      ),
+    );
+  }
 
-    final String transactionUrl =
-        (result.source as CardPaymentResponseSource).transactionUrl;
-
-    if (mounted) {
-      Navigator.push(
-        context,
-        MaterialPageRoute(
-            fullscreenDialog: true,
-            maintainState: false,
-            builder: (context) => ThreeDSWebView(
-                transactionUrl: transactionUrl,
-                callbackUrl: widget.config.callbackUrl,
-                on3dsDone: (String status, String message) async {
-                  if (status == PaymentStatus.paid.name) {
-                    result.status = PaymentStatus.paid;
-                  } else {
-                    result.status = PaymentStatus.failed;
-                    (result.source as CardPaymentResponseSource).message =
-                        message;
-                  }
-
-                  Navigator.pop(context);
-                  widget.onPaymentResult(result);
-                })),
-      );
+  void _on3dsDone(
+    PaymentResponse result,
+    String status,
+    String message,
+  ) {
+    if (status == PaymentStatus.paid.name) {
+      result.status = PaymentStatus.paid;
+    } else {
+      result.status = PaymentStatus.failed;
+      (result.source as CardPaymentResponseSource).message = message;
```

Comment:
We should check for
`if (result.source is CardPaymentResponseSource) {
result.source.message = message;
}`

Reply:
this is the response from moyasar gateway the made it dynamic